File: LeetCode_Algorithm/224(Basic Calculator)/224.js
var calculate = function(s) {
  // 136ms 12% 52.7MB 12%
  let stack = [];
  for (let c of s) {
    if (c == ' ') {
      continue
    } else if (c == ')') {
      let temp = []
      while (stack.length && stack[stack.length-1] != '(') {
        temp.push(stack.pop())
      }
      stack.pop()
      temp.reverse()
      stack.push(cal(temp))
    } else if (c == '+' || c == '-' || c == '(') {
      stack.push(c)
    } else {
      if (Number.isInteger(stack[stack.length-1])) {
        m = stack.pop()
        stack.push(10 * m + Number(c))
      } else {
        stack.push(Number(c))
      }
    }
  }  
  return cal(stack)
};

cal = (temp) => {
  if (temp[0] == '-') {
    ret = -temp[1]; temp.shift(); temp.shift()
  } else {
    ret = temp.shift()
  }
  while (temp.length) {
    m = temp.shift()
    if (m == '+') {
      ret += temp.shift()
    } else {
      ret -= temp.shift()
    }
  }
  return ret
}

console.log(calculate("2147483647"))

File: LeetCode_Algorithm/224(Basic Calculator)/224.py
class Solution:
    def calculate(self, s: str) -> int:
        # sol1 138ms 27% 18MB 26%
        stack = []
        temp = []
        for c in s:
            if c == ' ':
                continue
            elif c == ')':
                while stack and stack[-1] != '(':
                    a = stack.pop()
                    temp.append(a)
                stack.pop(); temp = temp[::-1]
                stack.append(self.cal(temp))
            elif c.isdigit():
                if stack and isinstance(stack[-1], int): 
                    n = stack.pop()
                    stack.append(10 * n + int(c))
                else:
                    stack.append(int(c))
            else:
                stack.append(c)
        return self.cal(stack)

    def cal(self, stack):
        if isinstance(stack[0], int):
            ret = stack[0]; stack.pop(0)
        else:
            stack.pop(0)
            n = stack.pop(0)
            ret = -n
        while stack:
            m = stack.pop(0)
            if m == '+':
                n = stack.pop(0)
                ret += n
            elif m == '-':
                if stack:
                    n = stack.pop(0)
                    ret -= n
                else:
                    return -1 * ret
        return ret
                

print(Solution().calculate("1-(     -2)"))
print(Solution().calculate("1-(-2)"))
print(Solution().calculate("-1-2+3"))



